packages/fuel-indexer-schema/src/db/queries.rs

```diff
@@ -0,0 +1,305 @@
+    pub namespace_identifier: String,
+    pub entity_name: String,
+    pub filters: Vec<EntityFilter>,
+}
+
+impl UserQuery {
+    pub fn to_sql(&mut self, db_type: &DbType) -> String {
+        // Different database solutions have unique ways of
+        // constructing JSON-formatted queries and results.
+        match db_type {
+            DbType::Postgres => {
+                let elements = self.parse_query_elements(db_type);
+
+                let elements_string = elements.join("");
+
+                let sorted_joins = self.get_topologically_sorted_joins();
+
+                let mut last_seen_primary_key_table = "".to_string();
+                let mut joins: Vec<String> = Vec::new();
+
+                for sj in sorted_joins {
+                    if sj.primary_key_table == last_seen_primary_key_table {
+                        if let Some(elem) = joins.last_mut() {
+                            let join_condition = format!(
+                                "{}.{} = {}.{}",
+                                sj.referencing_key_table,
+                                sj.referencing_key_col,
+                                sj.primary_key_table,
+                                sj.primary_key_col
+                            );
+                            *elem = format!("{elem} AND {join_condition}")
+                        }
+                    } else {
+                        joins.push(format!(
+                            "INNER JOIN {} ON {}.{} = {}.{}",
+                            sj.primary_key_table,
+                            sj.referencing_key_table,
+                            sj.referencing_key_col,
+                            sj.primary_key_table,
+                            sj.primary_key_col
+                        ));
+                        last_seen_primary_key_table = sj.primary_key_table;
+                    }
+                }
+
+                let mut query = format!(
+                    "SELECT json_build_object({}) FROM {}.{}",
+                    elements_string, self.namespace_identifier, self.entity_name,
+                );
+
+                if !joins.is_empty() {
+                    query = format!("{query} {}", joins.join(" "))
+                }
+
+                if !self.filters.is_empty() {
+                    let filter_str = self
+                        .filters
+                        .iter()
+                        .flat_map(|qf| {
+                            qf.filters.iter().map(|f| {
+                                f.to_sql(qf.fully_qualified_table_name.clone(), db_type)
+                            })
+                        })
+                        .collect::<Vec<String>>()
+                        .join(" AND ");
+                    query = format!("{query} WHERE {filter_str}");
+                }
+
+                println!("{query}");
+                query
+            }
+        }
+    }
+
+    fn parse_query_elements(&self, db_type: &DbType) -> Vec<String> {
```

Comment:
It would be helpful when you add rust-doc comments to explain what strategy you're using to parse these query elements. The more info the better